oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +241,336 @@ def all(self, length, keyword):
+
+    @defer.inlineCallbacks
+    def tcp_connect(self, socket):
+        cached_value = yield self.lookup('tcp_connect', socket)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        socket_info = {
+            'status': None,
+            'failure': None
+        }
+
+        ip_address, port = socket.split(":")
+        try:
+            point = TCP4ClientEndpoint(reactor, ip_address, int(port))
+            yield point.connect(TCPConnectFactory())
+            socket_info['status'] = True
+        except TimeoutError:
+            socket_info['status'] = False
+            socket_info['failure'] = 'generic_timeout_error'
+        except ConnectionRefusedError:
+            socket_info['status'] = False
+            socket_info['failure'] = 'connection_refused_error'
+        except:
+            socket_info['status'] = False
+            socket_info['failure'] = 'unknown_error'
+        yield self.cache_value('tcp_connect', socket, socket_info)
+        defer.returnValue(socket_info)
+
+    @defer.inlineCallbacks
+    def dns_consistency(self, hostname):
+        cached_value = yield self.lookup('dns_consistency', hostname)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        dns_info = {
+            'addrs': [],
+            'failure': None
+        }
+
+        try:
+            records = yield dns_client.lookupAddress(hostname)
+            answers = records[0]
+            for answer in answers:
+                if answer.type is dns.A:
+                    dns_info['addrs'].append(answer.payload.dottedQuad())
+                elif answer.type is dns.CNAME:
+                    dns_info['addrs'].append(answer.payload.name.name)
+        except DNSNameError:
+            dns_info['failure'] = 'dns_name_error'
+        except DNSServerError:
+            dns_info['failure'] = 'dns_server_failure'
+        except:
+            dns_info['failure'] = 'unknown_error'
+
+        yield self.cache_value('dns_consistency', hostname, dns_info)
+        defer.returnValue(dns_info)
+
+
+web_connectivity_cache = WebConnectivityCache()
+
+class WebConnectivity(RequestHandler):
+    @defer.inlineCallbacks
+    def control_measurement(self, http_url, socket_list):
+        hostname = urlparse(http_url).netloc
+        dl = [
+            web_connectivity_cache.http_request(http_url),
+            web_connectivity_cache.dns_consistency(hostname)
+        ]
+        for socket in socket_list:
+            dl.append(web_connectivity_cache.tcp_connect(socket))
+        responses = yield defer.DeferredList(dl)
+        http_request = responses[0][1]
+        dns = responses[1][1]
+        tcp_connect = {}
+        for idx, response in enumerate(responses[2:]):
+            tcp_connect[socket_list[idx]] = response[1]
+        self.finish({
+            'http_request': http_request,
+            'tcp_connect': tcp_connect,
+            'dns': dns
+        })
```

Comment:
I can confirm that passing invalid input now yields an HTTP error after d741df28a0c7b61a!